File: helper_func.py
#!/usr/bin/env python3
# Created on July 30, 2020
# This file holds a number of
# helper functions used throughout
# the project

import sys
import os.path
from os import path
import json


"""
Given the path to a directory, returns a list containing
the files in that directory 
"""
def read_directory(dir_name):
	files = []
	for filename in os.listdir(dir_name):
		filename = path.join(dir_name, filename)
		if path.isfile(filename):
			last_slash_index = filename.rfind('/')	
			filename = filename[last_slash_index + 2:] 
			files.append(filename)
	return files


"""
Given a filename and a row number,
confirms that the entered row is a 
valid row number. Returns the contents 
of the note as a list 
"""
def check_row(filename, row):
	if not row.isnumeric():
		print('You have entered in a faulty row number. Please choose an integer value.')
		return
	row = float(row)	
	if not row.is_integer():
		print('You have entered in a faulty row number. Please choose an integer value.')
		return
	row = int(row)
	memories = []
	with open(filename, 'r') as f:
		memories = f.readlines()
	if row <= 0 or row > len(memories):
		print('You have entered in a faulty row number. Please try again.')
		return
	return memories


"""
Given the quicknote_cache file, changes the
default file held in quicknote_cache
"""
def write_to_data_file(data_file, filename):
	data = {}
	data['current_note'] = filename
	with open(data_file, 'w') as f:
		json.dump(data, f)
	

"""
Given a prompt, captures the user's
answer as either 'y' or 'n'
"""
def request_user_permission(prompt):
	decision = input(prompt)
	while (decision != 'y') and (decision != 'n'):
		decision = input(prompt)
	return decision


"""
Given a set of arguments,
parses them and returns the
found word
"""
def parse_unary_args(args):
	if type(args) == str:
		return args.strip()
	if len(args) == 1:
		return args[0].strip()
	word = ''
	for arg in args:
		word += arg + ' '
	return word.strip()


"""
Given a set of arguments, parses
the words to the left and right of
the '/'
"""
def parse_binary_args(args):
	if len(args) == 1 or not '/' in args:
		print('You haven\'t entered enough information. Please try again.')
		return
	old_name = ''
	new_name = ''
	counter = 0
	while (not counter >= len(args)) and (not args[counter] == '/'):
		old_name += args[counter] + ' '
		counter += 1 		
	old_name = old_name.strip()
	counter += 1
	while not counter == len(args):
		new_name += args[counter] + ' '
		counter += 1 		
	new_name = new_name.strip()
	if len(new_name) == 0:
		print('You haven\'t entered enough information. Please try again.')
		return	
	return (old_name, new_name)


"""
Given a dot_name, a filename, and the file_list,
makes a new note with that name and
adds the name to file_list
"""
def make_note(dot_name, filename):
	if path.isfile(dot_name):
		print('A note with this name already exists. Please choose a different name, delete the other note, or rename the other note.')
		# return 0 to signify failure
		return 0
	else:
		open(dot_name, 'w+').close()

File: C_implementation/helper_func.c
#include <stdio.h>
#include <string.h>
#include <dirent.h>
#include <unistd.h>
#include "helper_func.h"

/*
 * METHODS 
 *
 * To Do:
 * readDirectory() --- requires array completion
 * checkRow() --- requires array completion
 * writeToDataFile() --- requires JSON completion 
 * parseUnaryArgs()/parseBinaryArgs() --- strip extra whitespace (grouped because have same issue)
 *
 * Seemingly Complete:
 * getFileSize()
 * printDefaultError()
 * printDirectory()
 * requestUserPermission()
 * makeNote()
 * parseUnaryArgs()
 * parseBinaryArgs()
 * copyFile()
 * getNotePath()
 * getDataFile();
 */


void getDataFile(char* dataFile) {
	char* dirName = "background";
	// while json-c is being figured out, data_file will act as data_file.json
	char* data = "data_file";
	getNotePath(dirName, data, dataFile);
}


void getNotePath(char* dirName, char* noteName, char* notePath) {
	char* tilde = "~";	
	wordexp_t newNote;
	// expand tilde bc that is user specific
	wordexp(tilde, &newNote, 0);
	
	strcpy(notePath, newNote.we_wordv[0]);
	strcat(notePath, "/.liszt/");
	strcat(notePath, dirName);
	strcat(notePath, "/");
	strcat(notePath, noteName);

	wordfree(&newNote);	

}


void copyFile(char* firstFile, char* secondFile) {
	FILE* source, * target;
	source = fopen(firstFile, "r");
	target = fopen(secondFile, "w");
	char filechar;
	while ((filechar = fgetc(source)) != EOF) {
		fputc(filechar, target);
	}
	fclose(source);
	fclose(target);
}


int getFileSize(char* filename) {
	FILE* toRead;
	char newline;
	int numLines = 0;
	toRead = fopen(filename, "r");
	while ((newline = fgetc(toRead)) != EOF) {
		if (newline == '\n') numLines++;
	}
	fclose(toRead);
	return numLines;
}


int checkDefault(char* note) {
	char loweredNote[256];
	strcpy(loweredNote, note);
	int numChars = strlen(loweredNote);
	for (int i = 0; i < numChars; i++) {
		loweredNote[i] = tolower(loweredNote[i]);
	}
	if (strcmp(loweredNote, "default") == 0) {
		printf("Sorry. 'default' is off limits.\n");
		return -1;
	}
	return 0;
}


void readDirectory(char* dirName, char* files[], int* numFiles) {
	DIR* directory;
	struct dirent* dir;
	directory = opendir(dirName);
	if (directory) {
		while ((dir = readdir(directory)) != NULL) {
			char* filename = dir->d_name;
			if (dir->d_type == DT_REG) {
				files[*numFiles] = filename;	
				(*numFiles)++;
			}
		}
		closedir(directory);
	}
}


void printDirectory(char* dirName, char* shortName) {
	DIR* directory;
	struct dirent* dir;
	directory = opendir(dirName);
	if (directory) {
		int counter = 0; // for counting the number of files
		while ((dir = readdir(directory)) != NULL) {
			if (dir->d_type == DT_REG) {
				counter++;
			}
		}
		if (strcmp(shortName, " ") == 0) counter--; // decrement counter bc don't want to include default in the count
		if (!counter) {
			printf("You have no notes at the moment.\n");
			return;
		}
		printf("\033[1mFound %d%snotes\033[0m\n", counter, shortName);
		rewinddir(directory);
		int index = 1;
		while ((dir = readdir(directory)) != NULL) {
			if (dir->d_type == DT_REG && strcmp(dir->d_name, "default") != 0) {
				printf("\033[1m%d.\033[0m %s\n", index, dir->d_name);
				index++;
			}
		}
		closedir(directory);
	}
}  


// int checkRow(char* filename, char* charRow, char*** memories, int* numMemories) {
// 	int row = atoi(charRow);	
// 	if (row <= 0) {
//         	printf("You have entered in a faulty row number. Please choose an integer value.\n");
// 		printf("hint: %s was not valid!\n", charRow);
// 		return 0;
// 	}

// 	FILE* toRead = fopen(filename, "r");
// 	int i = 0;
// 	while (fgets(&memories[i], sizeof(memories[i]), toRead)) {
// 		printf("Size of Memory: %lu\n", sizeof(memories[i]));
// 		i++;
// 	}
// 	*numMemories = i;
// 	if (row > *numMemories) {
// 		printf("You have entered in a faulty row number.\n");
// 		return 0;
// 	}
// 	fclose(toRead);
	
// 	for (int i = 0; i < *numMemories; i++) {
// 		printf("%s\n", memories[i]);
// 	}

// 	return 1;
// }

void writeToDataFile(char* filename) {
	char dataFile[256];
	getDataFile(dataFile);
	// The below is a stand-in until json-c has been figured out
	FILE* toWrite;
	toWrite = fopen(dataFile, "w");
	fprintf(toWrite, "%s", filename);
	fclose(toWrite);
}


void requestUserPermission(char* prompt, char* decision) {
	while (strcmp(decision, "y") != 0 && strcmp(decision, "n") != 0) {
		printf("%s", prompt);
		scanf("%99s", decision);
	}
}
	

int parseUnaryArgs(char* word, char* args[], int numArgs) {
	if (numArgs == 2) return 1; // just exit if it's a single word
	// start at 2 to avoid program invocation and command (argv[0] and argv[1])
	if (numArgs > 2) strcpy(word, args[2]);
	if (numArgs >= 4) strcat(word, " ");
	for (int i = 3; i < numArgs; i++) {
		strcat(word, args[i]);	
		if (numArgs > i + 1) strcat(word, " ");
	}
	return 0;
}


int parseBinaryArgs(char* first, char* second, char* args[], int numArgs) {
	// start at 2 to avoid program invocation and command (argv[0] and argv[1])
	if (numArgs > 2) strcpy(first, args[2]);
	if (numArgs >= 4 && strcmp(args[3], "/") != 0) strcat(first, " ");
	// confirm user entered a first word
	if (strcmp(args[2], "/") == 0) {
		printf("You have not entered enough information. Please try again.\n");
		return -1;
	}
	int counter = 3;
	while (counter < numArgs && strcmp(args[counter], "/") != 0) {
		strcat(first, args[counter]);	
		// trying to strip the last space from the first arg... leading to some issues with finding the file
		if (numArgs > counter + 2 && strcmp(args[counter + 1], "/") != 0) strcat(first, " ");
		counter++;
	} 
	counter++;
	if (counter >= numArgs) {
		printf("You have not entered enough information. Please try again.\n");
		return -1;
	} else strcpy(second, args[counter]);
	if (numArgs > counter + 1) strcat(second, " ");
	counter++;
	while (counter < numArgs) {
		strcat(second, args[counter]);
		if (numArgs > counter + 1) strcat(second, " ");
		counter++;
	}	
	return 0;
}


int makeNote(char* filePath) {
	if (access(filePath, F_OK) != -1) {
		printf("A note with this name already exists. Please choose a different name, delete the other note, or rename the other note.\n");
		return -1;
	}
	FILE* toCreate;
	toCreate = fopen(filePath, "w");
	fclose(toCreate);
	return 0;
}

